modules/cert-manager-acme-dns01-explicit-azure.adoc

```diff
@@ -0,0 +1,163 @@
+
+You can use {cert-manager-operator} to set up an ACME issuer to solve DNS-01 challenges by using explicit credentials on Microsoft Azure. This procedure uses _Let's Encrypt_ as the ACME CA server and shows how to solve DNS-01 challenges with Azure DNS.
+
+.Prerequisites
+
+* You have access to the {product-title} cluster as a user with the `cluster-admin` role.
+* You have set up an IAM role for Azure DNS. For more information, see link:https://cert-manager.io/docs/configuration/acme/dns01/azuredns/[Azure DNS] in the upstream cert-manager documentation.
++
+[NOTE]
+====
+You can follow this procedure for an {product-title} cluster that is not running on Microsoft Azure.
+====
+
+.Procedure
+
+. Optional: Override the nameserver settings for the DNS-01 self check:
++
+This step is required only when the target public-hosted zone overlaps with the cluster's default private-hosted zone.
+
+.. Edit the `CertManager` resource by running the following command:
++
+[source,terminal]
+----
+$ oc edit certmanager cluster
+----
+
+.. Add a `spec.controllerConfig` section with the following override arguments:
++
+[source,yaml]
+----
+apiVersion: operator.openshift.io/v1alpha1
+kind: CertManager
+metadata:
+  name: cluster
+  ...
+spec:
+  ...
+  controllerConfig:                                <1>
+    overrideArgs:
+      - '--dns01-recursive-nameservers-only'       <2>
+----
+<1> Add the `spec.controllerConfig` section.
+<2> Specify to only use recursive nameservers instead of checking the authoritative nameservers associated with that domain.
+
+.. Save the file to apply the changes.
+
+. Optional: Create a namespace for the issuer.
+
+.. Create a YAML file that defines a `Namespace` object:
++
+.Example `namespace.yaml` file
+[source,yaml]
+----
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: my-issuer-namespace <1>
+----
+<1> Specify the namespace for the issuer.
+
+.. Create the `Namespace` object by running the following command:
++
+[source,terminal]
+----
+$ oc create -f namespace.yaml
+----
+
+. Create a secret to store your Azure credentials in by running the following command:
++
+[source,terminal]
+----
+$ oc create secret generic azure-secret --from-literal=azureSecretAccessKey=<azure_secret_access_key> \ <1>
+    -n my-issuer-namespace
+----
+<1> Replace `<azure_secret_access_key>` with your Azure secret access key.
+
+. Create an issuer.
+
+.. Create a YAML file that defines the `Issuer` object:
++
+.Example `issuer.yaml` file
+[source,yaml]
+----
+apiVersion: cert-manager.io/v1
+kind: Issuer
+metadata:
+  name: issuer-acme-dns01-azuredns   <1>
+spec:
+  acme:
+    preferredChain: ""
+    privateKeySecretRef:
+      name: <secret_private_key> <2>
+    server: https://acme-staging-v02.api.letsencrypt.org/directory <3>
+    solvers:
+    - dns01:
+        azureDNS:
+          clientID: <azure_client_id> <4>
+          clientSecretSecretRef:
+            name: azuredns-config <5>
+            key: client-secret <6>
```

Comment:
> azure-secret --from-literal=azureSecretAccessKey=

Incorrect. YAML is azuredns-config but command is azure-secret. YAML is client-secret but command is azureSecretAccessKey.

Reply:
I mean the secret name and key name under clientSecretSecretRef YAML should be same as the command `oc create secret generic SECRET_NAME --from-literal=KEY_NAME=<azure_secret_access_key> ...`

Reply:
@xingxingxia Sorry, I didn't get this change. Is this expected? I changed YAML parameters according to the command
```clientSecretSecretRef:
            name: <secret_name> <5>
            key: <azure_secret_access_key> <6>```

Reply:
@xenolinux 
> @xingxingxia Sorry, I didn't get this change.

I mean: if the command looks like `oc create secret generic XXXX --from-literal=YYYY=<azure_secret_access_key>`, then under `Create a YAML file`, it should be:
```
          clientSecretSecretRef:
            name: XXXX
            key: YYYY
```
I.e. under `clientSecretSecretRef`, `name` refers to the secret name XXXX, `key` refers to the key **NAME** YYYY instead of the key value.
Your current version needs one more minor update:
Change
```
<6> Replace `<azure_secret_access_key>` with the client secret key.
```
To:
```
<6> Replace `<azure_secret_access_key_name>` with the client secret key name.
```